File: src/javascript/allowance.js
require('dotenv').config(); // Load environment variables
const { Web3 } = require('web3'); // Web3 setup

// Load environment variables
const SEPOLIA_URL = process.env.Sepolia_Url; // Sepolia network URL
const FLASHLOAN_ARBITRAGE_CONTRACT = process.env.Flashloan_Arbitrage; // Flashloan Arbitrage Smart Contract
const DEX_ADDRESS = process.env.Dex; // DEX address
const WALLET_ADDRESS = process.env.Wallet; // Wallet address

// ERC20 ABI for checking allowance
const ERC20_ABI = [
    {
        "constant": true,
        "inputs": [
            { "internalType": "address", "name": "_owner", "type": "address" },
            { "internalType": "address", "name": "_spender", "type": "address" }
        ],
        "name": "allowance",
        "outputs": [{ "internalType": "uint256", "name": "", "type": "uint256" }],
        "stateMutability": "view",
        "type": "function"
    }
];

// Connect to the Sepolia network
const web3 = new Web3(SEPOLIA_URL);

// Function to check allowance
async function checkAllowance(tokenAddress, spenderAddress) {
    try {
        const tokenContract = new web3.eth.Contract(ERC20_ABI, tokenAddress);
        const allowance = await tokenContract.methods.allowance(FLASHLOAN_ARBITRAGE_CONTRACT, spenderAddress).call();
        console.log(`Allowance for spender ${spenderAddress} from ${FLASHLOAN_ARBITRAGE_CONTRACT} for token ${tokenAddress}: ${allowance}`);
    } catch (error) {
        console.error("Error checking allowance:", error.message);
    }
}

// Example usage: Check USDC and DAI allowance for the DEX
const USDC_ADDRESS = process.env.Sepolia_USDC;
const DAI_ADDRESS = process.env.Sepolia_DAI;

checkAllowance(USDC_ADDRESS, DEX_ADDRESS);
checkAllowance(DAI_ADDRESS, DEX_ADDRESS);

File: src/python/allowance.py
import os
from dotenv import load_dotenv
from web3 import Web3

# Load environment variables
load_dotenv()
SEPOLIA_URL = os.getenv("Sepolia_Url")
FLASHLOAN_ARBITRAGE_CONTRACT = os.getenv("Flashloan_Arbitrage")
DEX_ADDRESS = os.getenv("Dex")
USDC_ADDRESS = os.getenv("Sepolia_USDC")
DAI_ADDRESS = os.getenv("Sepolia_DAI")

# Connect to Sepolia network
web3 = Web3(Web3.HTTPProvider(SEPOLIA_URL))
if not web3.is_connected():
    print("Failed to connect to Sepolia network")
    exit()

# ERC20 ABI for allowance function
ERC20_ABI = [
    {
        "constant": True,
        "inputs": [
            {"name": "_owner", "type": "address"},
            {"name": "_spender", "type": "address"}
        ],
        "name": "allowance",
        "outputs": [{"name": "", "type": "uint256"}],
        "type": "function"
    }
]

def check_allowance(token_address, spender_address):
    try:
        # Load token contract
        token_contract = web3.eth.contract(address=token_address, abi=ERC20_ABI)

        # Call allowance function
        allowance = token_contract.functions.allowance(
            FLASHLOAN_ARBITRAGE_CONTRACT, spender_address
        ).call()
        print(f"Allowance for spender {spender_address} on token {token_address}: {allowance}")
    except Exception as e:
        print(f"Error in checking allowance: {e}")

# Check allowances for USDC and DAI
check_allowance(USDC_ADDRESS, DEX_ADDRESS)
check_allowance(DAI_ADDRESS, DEX_ADDRESS)
